runtime/vm/createramclass.cpp

```diff
@@ -1529,17 +1529,33 @@ loadSuperClassAndInterfaces(J9VMThread *vmThread, J9ClassLoader *classLoader, J9
 				return FALSE;
 			}
 
+			UDATA superclassModifiers = superclass->romClass->modifiers;
 			/* ensure that the superclass isn't an interface or final */
-			if ((superclass->romClass->modifiers & J9_JAVA_FINAL) != 0) {
+			if (J9_ARE_ALL_BITS_SET(superclassModifiers, J9AccFinal)) {
 				Trc_VM_CreateRAMClassFromROMClass_superclassIsFinal(vmThread, superclass);
 				setCurrentExceptionForBadClass(vmThread, superclassName, J9VMCONSTANTPOOL_JAVALANGVERIFYERROR);
 				return FALSE;
 			}
-			if ((superclass->romClass->modifiers & J9_JAVA_INTERFACE) != 0) {
+			if (J9_ARE_ALL_BITS_SET(superclassModifiers, J9AccInterface)) {
 				Trc_VM_CreateRAMClassFromROMClass_superclassIsInterface(vmThread, superclass);
 				setCurrentExceptionForBadClass(vmThread, superclassName, J9VMCONSTANTPOOL_JAVALANGINCOMPATIBLECLASSCHANGEERROR);
 				return FALSE;
 			}
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+			/* ensure that the superclass isn't a value type */
+			if (J9_ARE_ALL_BITS_SET(superclassModifiers, J9AccValueType)) {
+				Trc_VM_CreateRAMClassFromROMClass_superclassIsValueType(vmThread, superclass);
+				setCurrentExceptionNLS(vmThread, J9VMCONSTANTPOOL_JAVALANGINCOMPATIBLECLASSCHANGEERROR, J9NLS_BCV_SUPERCLASS_IS_VALUE_TYPE);
+				return FALSE;
+			}
+
+			/* ensure that value type classes inherit from j.l.Object */
+			if (J9_ARE_ALL_BITS_SET(romClass->modifiers, J9AccValueType) && !(superclass == J9VMJAVALANGOBJECT_OR_NULL(vm))) {
```

Comment:
`!(superclass == J9VMJAVALANGOBJECT_OR_NULL(vm))` can just be `(J9VMJAVALANGOBJECT_OR_NULL(vm) != superclass)`

Also, remember to always put the rvalue on the left side of the equality checks